x-pack/plugins/lens/public/editor_frame_plugin/editor_frame/expression_helpers.ts

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Ast, fromExpression } from '@kbn/interpreter/common';
+import { Visualization, Datasource, DatasourcePublicAPI } from '../../types';
+
+export function buildExpression(
+  visualization: Visualization,
+  visualizationState: unknown,
+  datasource: Datasource,
+  datasourceState: unknown,
+  datasourcePublicAPI: DatasourcePublicAPI
+): Ast | undefined {
+  const datasourceExpression = datasource.toExpression(datasourceState);
+  const visualizationExpression = visualization.toExpression(
+    visualizationState,
+    datasourcePublicAPI
+  );
```

Comment:
I think the assumption that both datasource and visualization will always return a useful expression is the cause of the problem. I think we've seen that in Canvas, using sample data to replace a real data set is confusing to users. Visualizations without data might be able to render some kind of frame, but are otherwise meaningless. So here's a proposal:

1. Allow datasources and visualizations to not provide an expression. Could use either `null` or `""` (empty string) to represent this.
2. If the datasource expression is not provided, I would also not render the visualization. This is mostly because of the suggestion system, where we might switch the visualization type based on the selected data.

Reply:
Makes sense, I implemented it like this. If one of the two doesn't provide an expression, nothing is rendered